include/swift/Reflection/ReflectionContext.h

```diff
@@ -801,7 +801,50 @@ class ReflectionContext
     }
 
     case RecordKind::MultiPayloadEnum: {
-      // TODO: Support multipayload enums
+      // Collect basic statistics about the enum
+      unsigned long PayloadCaseCount = 0;
+      unsigned long NonPayloadCaseCount = 0;
+      unsigned long PayloadSize = 0;
+      for (auto Field : Fields) {
+        if (Field.TR != 0) {
+          PayloadCaseCount += 1;
+          if (Field.TI.getSize() > PayloadSize) {
+            PayloadSize = Field.TI.getSize();
+          }
+        } else {
+          NonPayloadCaseCount += 1;
+        }
+      }
+      if (EnumSize > PayloadSize) {
+        // If the compiler laid this out with a separate tag, use that.
+        unsigned tag = 0;
+        auto TagSize = EnumSize - PayloadSize;
+        auto TagAddress = remote::RemoteAddress(EnumAddress.getAddressData() + PayloadSize);
+        if (!getReader().readInteger(TagAddress, TagSize, &tag)
+           || tag >= Fields.size()) {
+          return false;
+        }
+        if (tag < PayloadCaseCount) {
+          *CaseIndex = tag;
+          return true;
+        }
+        auto PayloadTagSize = std::min(PayloadSize, static_cast<size_t>(4));
+        // Treat the tag as a page selector; payload carries the offset within the page
+        auto Page = tag - PayloadCaseCount;
+        // Zero for 32-bit because we'll never have more than one page
+        auto PageSize = PayloadTagSize >= 4 ? 0 : 1 << (PayloadSize * 8U);
+        auto PageStart = Page * PageSize;
+        unsigned PayloadTag;
+        if (!getReader().readInteger(EnumAddress, PayloadTagSize, &PayloadTag)) {
+          return false;
+        }
+        *CaseIndex = PageStart + PayloadTag + PayloadCaseCount;
+        return true;
+      } else {
+        // XXX TODO: If the payloads have common spare bits (e.g., all pointers)
+        // then use those to decode the case.
```

Comment:
I think only the compiler really knows if spare bits are used or not. What I suggest is having IRGen emit a new kind of descriptor for fixed-size multi-payload enums that use spare bits. The descriptor can tell you which bits in the payload to use for the discriminator. If no such descriptor is found, you can fall back to using a tag here.

Reply:
> If no such descriptor is found, you can fall back to using a tag here.

Hmmm....  This doesn't seem quite right.  If there's no such descriptor, that could just mean that this was from an old compiler that didn't include the data.  I think the logic should be:
* If enum is bigger than the max payload size, then use a tag.  (This is the existing test above.)
* If there's a descriptor giving spare bit info, use that.
* Otherwise, return `false` -- this is a multi-payload enum in an unknown format (possibly because it came from a newer compiler)

Reply:
You're absolutely right. As we discussed offline, the compiler already emits builtin descriptors for fixed-size multi-payload enums, so the existence of such a descriptor should be plumbed through from type lowering to here.

Reply:
I have a start of that in another branch here; I hope to be able to submit by the end of the week.